nixos/modules/config/supported-hashes.py

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python3
+import sys
+
+
+# https://github.com/besser82/libxcrypt/blob/v4.4.28/lib/hashes.conf#L41
+good_schemes = [
+    "y",  # yescrypt
+    "gy",  # gost_yescrypt
+    "7",  # scrypt
+    "2b",  # bcrypt
+    "2y",  # bcrypt_y
+    "2a",  # bcrypt_a
+    "6",  # sha512crypt
+]
+
+
+def is_deprecated(hash: str):
+    try:
+        _, scheme, _ = hash.split("$", maxsplit=2)
+    except ValueError:
+        # descrypt
+        return True
+
+    return scheme not in good_schemes
+
+
+users = set()
+try:
+    with open("/etc/shadow", "r") as fd:
+        for line in fd:
+            try:
+                user, hash, _ = line.split(":", maxsplit=2)
+            except ValueError:
+                # malformed?
+                continue
+
+            if hash == "!":
+                # no password set
+                continue
+
+            if is_deprecated(hash):
+                users.add(user)
+
+except PermissionError as ex:
+    print(ex, file=sys.stderr)
+
+if not users:
+    sys.exit(0)
+
+print(
+    """
+    WARNING: The following users are relying on password hashes that will
+    be removed in NixOS 23.05. They should be renewed as soon as possible!""",
+    file=sys.stderr,
+)
```

Comment:
Can we abort the switching in a next deprecation step?

Reply:
I also thought this was a good idea, but realized it's an obvious bad idea when doing `nixos-rebuild boot` into e.g. 23.05 -- the system would just fail to boot with that message, which... I don't know if we want that. Maybe we do? Not sure.

Reply:
We can harden the validation for `hashedPassword` using `strMatching`, but we don't have an equally good solution for mutable passwords.